### ISSUE TITLE
[PHP] Unsafe updating of variation data

### ISSUE THREAD

**Issue by saulfautley** [woocommerce/woocommerce#24956]
**Describe the bug**
If you instantiate a new `WC_Product_Variation` object with the ID of any `variable` product, the `variable` product's title is **_immediately_** updated in the database to the global `$post->post_title` (or blank if there is no global `$post`).

This seems like a pretty critical issue as someone could easily be developing or debugging and instantiate `WC_Product_Variation` instead of `WC_Product_Variable` by mistake one time and BAM, the `variable` product's title is now "Home page" or something. In fact it's not just limited to the title as `post_excerpt` is also updated on `read()`. The worst part is that you probably won't even notice until you actually come across that product again.

I just encountered this in a pretty bad way on a production site with the WooCommerce Wishlists plugin that has loads of wishlists with saved product variations stored as serialized `WC_Product_Variation` objects, but with the IDs set to the ID of the parent `WC_Product_Variable` for some unholy reason. Viewing a public wishlist would immediately set all of the product titles to the name of the wishlist! What a shitshow that was to debug...

Of course, the only time this issue isn't an _issue_ is when the global `$post` is of the same `variable` product for which the ID is being passed (i.e. on its own edit or view page).

**To Reproduce**
```php
global $post;
$post->post_title = 'foo';
new WC_Product_Variation(100); // ID of any variable product
echo wc_get_product(100)->get_title(); // foo (this is now permanent!)
```

**Expected behavior**
`WC_Product_Variation_Data_Store_CPT::read` should at the very least not update database values when `$post_object->post_parent` is 0.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Environment**
WooCommerce 3.7.1

**Comment by claudiosanches** [comment]
This may happens because of code to keep backwards compatibility:

https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-variation-data-store-cpt.php#L79-L105

For now the best solution to avoid problems like that is to improve this check:

https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-variation-data-store-cpt.php#L51-L53

It should check only for `product_variation`.

Still this is a misuse of `WC_Product_Variation` too, since none of the product type classes should be instantiated directly, everything should pass by the `WC_Product_Factory` class, where is possible to proper map each product type to the specific class, also it allows 3rd party plugins to override WooCommerce's product type classes, as you can see in:

https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-product-factory.php#L52-L68

So this also need to be fixed on WooCommerce Wishlists plugin, or any other code.

**Comment by claudiosanches** [comment]
Thank you for reporting this issue!

I'm closing the issue as it already has a PR, please continue the discussion over at #25030.

**Comment by claudiosanches** [comment]
> WC_Product_Variation_Data_Store_CPT::read should at the very least not update database values when $post_object->post_parent is 0.

Since some changes are to better allow import products, so it's better to not change this code, because as I said, you should not directly instantiate any product type class, since those can be changed by 3rd plugin plugins.

Instead of changing it, I'm including a new helper in #25031, I hope this will help 3rd party developers.

**Comment by saulfautley** [comment]
Fair enough. Looks like a good update.

However I don't think it would cause any problems if we just added something like this
```php
if (!$product->get_parent_id()) {
    $product->set_object_read(true);
    return;
}
```
after
https://github.com/woocommerce/woocommerce/blob/0f28a6191076a57ef6576d9196a863bc79499971/includes/data-stores/class-wc-product-variation-data-store-cpt.php#L76

Everything below that looks like it will fail when parent_id is 0 anyway, which is the cause of this issue in the first place.

**Comment by claudiosanches** [comment]
@saulfautley I'll write a Pull Request for it.

**Comment by saulfautley** [comment]
@claudiosanches you are a powerhouse man 😄

**Comment by claudiosanches** [comment]
Solved by #25178
